libunwind/src/UnwindRegistersSave.S

```diff
@@ -897,8 +884,20 @@ DEFINE_LIBUNWIND_FUNCTION(__unw_getcontext)
   stp    d28,d29, [c0, #0x0e0]
   str    d30,     [c0, #0x0f0]
   str    d31,     [c0, #0x0f8]
+#ifdef _LIBUNWIND_HAS_CHERI_LIB_C18N
+  // Store the trusted stack pointer
+  stp    c0, c30, [csp, #-0x20]!
+  bl     dl_c18n_get_trusted_stk
```

Comment:
What happens when we return back to the calling compartment? Do we not end up referencing a trusted frame that no longer exists?

Reply:
The implementation of `dl_c18n_get_trusted_stk` already accounts for this and returns the trusted frame below the top one.

Reply:
Well that doesn't work if you have, say, something within [TCB] call setjmp?

Reply:
That seems surprising and fragile, no?

Reply:
Well, `setjmp` is marked as a trusted symbol so calling it anywhere won't result in a new frame being pushed. `setjmp` calling `dl_c18n_get_trusted_stk` will push a new frame which will be skipped over in the implementation of `dl_c18n_get_trusted_stk`. So this always works.

Reply:
Hm, it's not ideal having that as a requirement for the caller. In these contexts it's fine, but ideally it'd be general enough that you could ask for your own trusted frame or your caller's. Anything more than that is going to need the full complexity of unwinding involved to be useful so having to go through libunwind to do it is fine, but I could imagine both those options being useful.

To throw some ideas around, you could:
* Have dl_c18n_get_trusted_stack take a bool for whether to pop a frame (you'd have to have the caller also call dl_c18_is_trampoline on CLR unless it knows it's a trusted symbol)
* Have dl_c18n_get_trusted_stack take a pointer (i.e. CLR) which it then calls dl_c18n_is_trampoline on and proceeds like the above (with NULL being treated as not a trampoline, i.e. get the current frame, if dl_c18n_is_trampoline itself doesn't do that)

Reply:
Sure. I've implemented the second option and modified `dl_c18n_is_trampoline` to check for the tag on the CLR passed in.